include/deal.II/hp/refinement.h

```diff
@@ -159,36 +159,83 @@ namespace hp
+      const ComparatorRefine &             compare_refine = ComparatorRefine(),
+      const ComparatorCoarsen &compare_coarsen = ComparatorCoarsen());
+
+    /**
+     * Adapt the finite element on cells whose indicators meet a certain
+     * threshold relative to their value range.
      *
      * The threshold will be chosen for refined and coarsened cells
      * individually. For each class of cells, we determine the maximal and
-     * minimal values of the smoothness indicators and determine the threshold
-     * by linear interpolation between these limits. Parameters
-     * @p p_refine_fraction and @p p_refine_coarsen are used as interpolation
-     * factors, where `0` corresponds to the minimal and `1` to the maximal
-     * value. By default, mean values are considered as thresholds.
+     * minimal values of the indicators and determine the threshold by linear
+     * interpolation between these limits. Parameters @p p_refine_fraction and
+     * @p p_refine_coarsen are used as interpolation factors, where `0`
+     * corresponds to the minimal and `1` to the maximal value. By default, mean
+     * values are considered as thresholds.
      *
-     * We consider a cell for p-refinement if it is flagged for refinement and
-     * its smoothness indicator is larger than the corresponding threshold. The
-     * same applies for p-coarsening, but the cell's indicator must be lower
-     * than the threshold.
+     * We consider a cell for p-adaptivity if it is flagged correspondingly and
+     * its indicator successfully compares to the corresponding threshold. Let
+     * us be more specific on the default case: We consider a cell for
+     * p-refinement if it is flagged for refinement and its indicator is larger
+     * than the corresponding threshold. The same applies for p-coarsening, but
+     * the cell's indicator must be lower than the threshold. However, different
+     * comparison functors can be supplied via the parameters @p compare_refine
+     * and @p compare_coarsen to impose different decision criteria.
      *
-     * Each entry of the parameter @p smoothness_indicators needs to correspond
-     * to an active cell. Parameters @p p_refine_fraction and
-     * @p p_coarsen_fraction need to be in the interval $[0,1]$.
+     * Each entry of the parameter @p indicators needs to correspond to an active
+     * cell. Parameters @p p_refine_fraction and @p p_coarsen_fraction need to be
+     * in the interval $[0,1]$.
      *
      * @note Preceeding calls of Triangulation::prepare_for_coarsening_and_refinement()
      *   may change refine and coarsen flags, which will ultimately change the
      *   results of this function.
      */
-    template <int dim, typename Number, int spacedim>
+    template <int dim,
+              typename Number,
+              int spacedim,
+              typename ComparatorRefine  = std::greater<Number>,
+              typename ComparatorCoarsen = std::less<Number>>
```

Comment:
I was undecided whether to use std::functions or functions passed as templates for this task, and I am not sure which one of these approaches would be better for performance.

I am okay with using std::function if they are preferred in the library. I think we use them for callbacks, too. So maybe playing around with function via templates would be the task for a different pull request, involving some benchmarking. I will undo the changes to use std::function again.

Reply:
No, you need `std::less<double>()`. Note the parentheses -- you need to create an object.

Reply:
Sorry, I think I wasn't specific enough about my issue.

Our unit test `hp/p_adaptivity_flags.cc` complains as follows. With the removal of the comparator templates, we are no longer able to override default comparators like this:
```
hp::Refinement::p_adaptivity_from_reference(
  dh, criteria, references, std::less<double>(), std::greater<double>());
```
We now have to provide the templates for the function explicitly:
```
hp::Refinement::p_adaptivity_from_reference<dim, double>(
  dh, criteria, references, std::less<double>(), std::greater<double>());
```
Otherwise, we would get the above error. It appears that the compiler doesn't get that `double` is both the value type of `Vector<double>` and `std::less|greater<double>`. 

Anyways, I'd be fine with this drawback of explicitly providing template parameters.

Reply:
I see. That's because it tries (but fails) to deduce `Number` from `std::function<bool (Number, Number)>` because the argument is not in fact given as such. But you can rewrite the declaration to use `std::function<bool (typename identity<Number>::type, typename identity<Number>::type)>` and it should work. (Maybe with `const` and `&` for the arguments, depending on how you do it right now.

Reply:
Thanks you for that hint! I didn't know about it yet! This is a C++20 feature. In this case, we need to pick `type_identity` instead of `identity`.

I introduced an alias to make things more easily readable.